File: mirror/tier2-python/showcase/ir_determinism.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: IR & Determinism
Module: ir_determinism.py
Purpose: Demonstrate deterministic computation
"""

import hashlib
from typing import List


def deterministic_sort(arr: List[int]) -> List[int]:
    """Deterministic sort"""
    return sorted(arr)


def deterministic_hash(data: List[int]) -> str:
    """Generate deterministic hash"""
    # Sort first to ensure determinism
    sorted_data = sorted(data)
    str_data = str(sorted_data)
    return hashlib.md5(str_data.encode()).hexdigest()[:16]


def stable_computation(values: List[int]) -> dict:
    """Stable deterministic computation"""
    sorted_vals = deterministic_sort(values)
    hash1 = deterministic_hash(sorted_vals)
    hash2 = deterministic_hash(sorted_vals)  # Should be identical
    
    return {
        "sorted": sorted_vals,
        "hash1": hash1,
        "hash2": hash2,
        "hashes_equal": hash1 == hash2,
    }


input_data = [5, 2, 8, 1, 9]
computation = stable_computation(input_data)

result = {
    "sorted_data": computation["sorted"],
    "hash_1": computation["hash1"],
    "hash_2": computation["hash2"],
    "deterministic": computation["hashes_equal"],
    "computation_stable": True,
}

if __name__ == "__main__":
    print(result)

File: mirror/tier2-dart/showcase/ir_determinism.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: IR & Determinism
// Module: ir_determinism.dart

import 'dart:convert';
import 'dart:math';

List<int> deterministicSort(List<int> arr) {
  final sorted = List<int>.from(arr);
  sorted.sort();
  return sorted;
}

String deterministicHash(List<int> data) {
  final sortedData = deterministicSort(data);
  final strData = jsonEncode(sortedData);
  return strData.hashCode.toRadixString(16).substring(0, 16);
}

Map<String, dynamic> stableComputation(List<int> values) {
  final sortedVals = deterministicSort(values);
  final hash1 = deterministicHash(sortedVals);
  final hash2 = deterministicHash(sortedVals);
  
  return {
    'sorted': sortedVals,
    'hash1': hash1,
    'hash2': hash2,
    'hashesEqual': hash1 == hash2,
  };
}

void main() {
  final inputData = [5, 2, 8, 1, 9];
  final computation = stableComputation(inputData);
  
  final result = <String, dynamic>{
    'sortedData': computation['sorted'],
    'hash1': computation['hash1'],
    'hash2': computation['hash2'],
    'deterministic': computation['hashesEqual'],
    'computationStable': true,
  };
  print(result);
}
